### PR TITLE
Add bucksense Adapter

### PR DESCRIPTION
**Pull request by bkse-stefanodechicchis**
## Type of change

- [X ] New bidder adapter   

## Description of change
This is first release for Bucksense Adapter

- test parameters for validating bids
```
{
  bidder: 'bucksense',
  params: {
    placementId : 1000
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer (stefano.dechicchis@bucksense.com)
- [ X] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

**Comment by bretg** [comment]
docs PR https://github.com/prebid/prebid.github.io/pull/1284

File: modules/bucksenseBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  supportedMediaTypes: [BANNER],
+
+  /**
+   * Determines whether or not the given bid request is valid.
+   *
+   * @param {object} bid The bid to validate.
+   * @return boolean True if this is a valid bid, and false otherwise.
+  */
+  isBidRequestValid: function (bid) {
+    if (isNaN(bid)) {
```
Comment (jsnellbaker) [inline modules/bucksenseBidAdapter.js]: The `bid` param is an object at this stage of the code, so this if statement check will always return `true` and cause the bidRequest for your bidder to be considered invalid.

Can you review this logic and adjust it accordingly?

File: modules/bucksenseBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    for (let i = 0; i < len; i++) {
+      var bid = validBidRequests[i];
+
+      requests.push({
+        method: 'POST',
+        url: URL,
+        data: {
+          'pub_id': location.host,
+          'pl_id': bid.params.placementId,
+          'sys_href': encodeURI(location.href),
```
Comment (jsnellbaker) [inline modules/bucksenseBidAdapter.js]: If you haven't already looked at it, there are a set of data points you can use in your adapter to get information on the webpage loading the adapter.  Please see the link below and consider using them if they would fit your needs:
http://prebid.org/dev-docs/bidder-adaptor.html#referrers

File: modules/bucksenseBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  },
+
+  /**
+  * Register the user sync pixels which should be dropped after the auction.
+  *
+  * @param {SyncOptions} syncOptions Which user syncs are allowed?
+  * @param {ServerResponse[]} serverResponses List of server's responses.
+  * @return {UserSync[]} The user syncs which should be dropped.
+  */
+  getUserSyncs: function(syncOptions, serverResponses) {
```
Comment (jsnellbaker) [inline modules/bucksenseBidAdapter.js]: Is this code still a work-in-progress?

If not and you're not planning to support userSyncs (at least at this time) - you can just remove this function entirely from your adapter.

File: modules/bucksenseBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    if (bksdebug) console.log(WHOIS + ' onBidWon() - INPUT bid:', bid);
+  },
+
+  /**
+   * ...
+   *
+   * @param ...
+   * @return ...
+  */
+  onTimeout: function(timeoutData) {
```
Comment (jsnellbaker) [inline modules/bucksenseBidAdapter.js]: Similar to the getUserSyncs function - you can remove this function if you're not planning to use it.

File: modules/bucksenseBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    return syncs;
+  },
+
+  /**
+   * ...
+   *
+   * @param {object} bid The bid to validate.
+   * @return ...
+  */
+  onBidWon: function(bid) {
```
Comment (jsnellbaker) [inline modules/bucksenseBidAdapter.js]: Similar to the getUserSyncs function - you can remove this function if you're not planning to use it.

File: modules/bucksenseBidAdapter.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
+import { registerBidder } from 'src/adapters/bidderFactory';
+import { BANNER } from 'src/mediaTypes';
+
+const WHOIS = 'BKSHBID-008';
+const BIDDER_CODE = 'bucksense';
+const URL = 'https://prebid.bksn.se:445/prebid';
```
Comment (jsnellbaker) [inline modules/bucksenseBidAdapter.js]: Is this the correct endpoint URL for your adapter?

When I tried to test the adapter, the POST requests to this URL were resulting in a 301 Moved Permanently response.

File: modules/bucksenseBidAdapter.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
+import { registerBidder } from 'src/adapters/bidderFactory';
+import { BANNER } from 'src/mediaTypes';
```
[the remark this answers is not in the document]
Comment (bkse-stefanodechicchis) [reply modules/bucksenseBidAdapter.js]: fixed with new Push

File: modules/bucksenseBidAdapter.js (hunk 1 above)
Comment (bkse-stefanodechicchis) [reply modules/bucksenseBidAdapter.js]: fixed with new Push

File: modules/bucksenseBidAdapter.js (hunk 2 above)
Comment (bkse-stefanodechicchis) [reply modules/bucksenseBidAdapter.js]: Thanks for point to this but in this release we would like to keep using the location object if it is not a problem

File: modules/bucksenseBidAdapter.js (hunk 3 above)
Comment (bkse-stefanodechicchis) [reply modules/bucksenseBidAdapter.js]: function not needed on this release, we removed.

File: modules/bucksenseBidAdapter.js (hunk 4 above)
Comment (bkse-stefanodechicchis) [reply modules/bucksenseBidAdapter.js]: function not needed on this release, we removed.

File: modules/bucksenseBidAdapter.js (hunk 5 above)
Comment (bkse-stefanodechicchis) [reply modules/bucksenseBidAdapter.js]: function not needed on this release, we removed.

File: modules/bucksenseBidAdapter.js (hunk 6 above)
Comment (bkse-stefanodechicchis) [reply modules/bucksenseBidAdapter.js]: Yes request is doing a 301 but then got redirect to the script that return the BID. I have tested the prebid on the Hello World example. Let me know if you do not see the response at all from your test.

**Comment by jsnellbaker** [comment]
Hi @bkse-stefanodechicchis, Thanks for making the updates.

I retested the adapter and was now able to see the bid return and render successfully.

However, when I reran the coverage tests - the adapter is still below the ideal; it's at 72% currently.
Can you put together some additional unit tests to fill in the gaps?

Also can you update the various `console.log` statements in the adapter file to use the`utils.logInfo()` function instead?  This function would categorize/brand the messages in the console that they're part of Prebid and would be more ideal to use.

We should be good to merge once these last few items are addressed.  Thanks.

**Comment by bkse-stefanodechicchis** [comment]
Hi,
I'm struggling to increase the percentage of testing.
It is my first experience with Chai and this is not helping.
My specs file already test all the 4 function that are in the Adapter, so I
thought I would add more scenarios to check in each function, but if I add
2 new scenario, I don't see the % going up.
I appreciate any suggestions on what I'm doing wrong
Thanks

**Comment by bkse-stefanodechicchis** [comment]
I'm struggling to increase the percentage of testing.
Screen Shot 2019-05-03 at 4 12 47 PM

My specs file already test all the 4 functions that are in the Adapter, so I thought I would add more scenarios to check in each function, but if II added some more and I don't see the % going up, so I think I'm in the wrong direction.
I appreciate any suggestions/help on what I'm doing wrong
Thanks

**Comment by jsnellbaker** [comment]
Hi @bkse-stefanodechicchis 

Just to clarify, you are looking at your specific adapter's coverage rate correct?  It's inside the modules sub-folder in that initial screen.  See below for example:

Screen Shot 2019-05-06 at 8 25 44 AM

From what I can see (currently with what has been committed), the outstanding areas all relate to the `bksdebug` param.  

If you setup a pair of tests where that field is enabled (one for `requestBids` and one for `interpretResponse`) and swap out the `console.log` with `utils.logInfo` - you could create a stub of the `utils.logInfo` function and run your checks against that stub (ie check for the `called` property is `true` for example).  These new tests should help fill the gap.

**Comment by bkse-stefanodechicchis** [comment]
I just commit a new version, I removed the `console.log` and use `utils.logInfo`.
I'm not able to reach the 80% coverage I also removed the bksdebug param and added some more use case but the % just not moving up.

I look at the file from the `gulp view-coverage`
image
and I'm try to figure out why some IF or ELSE get a flag

**Comment by jsnellbaker** [comment]
Hi @bkse-stefanodechicchis I checked out your updates to the adapter test cases.  I'm seeing your adapter coverage at 94% now (see below):
Screen Shot 2019-05-07 at 10 15 17 AM

Are you not seeing the same thing?  If you're happy with how your adapter is handling certain scenarios (this is in reference to your last comment on the if/else statements in the `interpretResponse`), we should be okay to merge the PR.  

Please let me know when you have the chance.

**Comment by bkse-stefanodechicchis** [comment]
Hi, I get confuse I was looking at column "Branches" and the % was low there.
Anyway we are good in our side, so ready for the merge from our point of view.
Thanks

**Pull request merged**